release/train_tests/benchmark/recsys/torchrec_runner.py

```diff
@@ -0,0 +1,57 @@
+import logging
+import os
+
+import torch
+from torchrec.distributed import TrainPipelineSparseDist
+from torchrec.optim.keyed import CombinedOptimizer, KeyedOptimizerWrapper
+from torchrec.optim.optimizers import in_backward_optimizer_filter
+
+import ray.train
+import ray.train.torch
+
+from runner import TrainLoopRunner
+
+
+logger = logging.getLogger(__name__)
+
+
+class TorchRecRunner(TrainLoopRunner):
+    def setup(self):
+        device = ray.train.torch.get_device()
+
+        self.model = self.factory.get_model()
+
+        dense_optimizer = KeyedOptimizerWrapper(
+            dict(in_backward_optimizer_filter(self.model.named_parameters())),
+            lambda params: torch.optim.Adagrad(params, lr=15.0, eps=1e-8),
+        )
+        self.optimizer = CombinedOptimizer(
+            [self.model.fused_optimizer, dense_optimizer]
+        )
+        # lr_scheduler = LRPolicyScheduler(
+        #     optimizer, args.lr_warmup_steps, args.lr_decay_start, args.lr_decay_steps
+        # )
+
+        self.pipeline = TrainPipelineSparseDist(
+            self.model, self.optimizer, device, execute_all_batches=True
+        )
+
+    def train_step(self, train_dataloader):
+        self.pipeline.progress(train_dataloader)
```

Comment:
Torchrec does some things differently and didn't fit the old `TrainLoopRunner` mold.
1. Creates this `TrainPipelineSparseDist` that wraps the model and optimizer, so the user doesn't call `model(input)` or compute the loss function directly. This pipelines a few batches at a time (moving to GPU, doing collective communication, forward and backward pass).
2. It also doesn't call next on the data iterator directly, but rather does that in `pipeline.progress(dataloader)`.
3. Does *not* wrap the model in DDP, instead wraps it in a torchrec `DistributedModelParallel` utility class.

Reply:
So, I pulled out some methods that can be defined by subclasses. `setup`, `train_step`, validate_step`. Setup initializes all model related components that should be used in the training/validate steps.

Reply:
I found a different API in torchrec that allows calling next on the data iterator directly and does the forward/backward/optimizer steps explicitly.

However, there are are still differences between the 2, so I kept the overridable `Runner` skeleton.